aten/src/ATen/native/TensorFactories.cpp

```diff
@@ -727,5 +727,195 @@ Tensor tensor_cuda(ArrayRef<T> values, const TensorOptions& options) {
+  );
+
+  auto options = x.options().dtype(at::kLong);
+  THGenerator* generator = get_generator(nullptr);
+
+  if (weights.numel() == 0){
+    Tensor indices_n = at::arange({n}, options);
+    int split, begin, end;
+    if(2 * k < n){
+      split = n - k;
+      begin = n - k;
+      end = n;
+    } else {
+      split = k;
+      begin = 0;
+      end = k;
+    }
+
+    reservoir_generator_cpu(
+      indices_n.data<int64_t>(),
+      n,
+      split,
+      generator);
+
+      return x.index_select(
+        0,
+        indices_n.index_select(
+          0,
+          at::arange(begin, end, options)
+        )
+      );
+
+  } else {
+
+    AT_CHECK(
+      weights.device() == x.device(),
+      "The weights must share the same device as the inputs."
+    );
+
+    AT_CHECK(
+      n == weights.numel(),
+      "The weights must have the same number of elements as the input's first dimension."
+    );
+
+    AT_CHECK(
+      weights.dim() == 1,
+      "The weights must 1-dimensional."
+    );
+
+    AT_CHECK(
+      weights.nonzero().numel() >= k,
+      "Cannot have less non-zero weights than the number of samples."
+    );
+
+    AT_CHECK(
+      weights.min().item().toLong() >= 0,
+      "All the weights must be non-negative."
+    );
+
+    Tensor keys = at::empty({n}, weights.options());
+
+    AT_DISPATCH_FLOATING_TYPES(weights.scalar_type(), "generate keys", [&] {
+      generate_keys<scalar_t>(
+        keys.data<scalar_t>(),
+        weights.data<scalar_t>(),
+        n,
+        generator);
+    });
+
+    return x.index_select(0, std::get<1>(keys.topk(k)));
+  }
+
+}
+
+Tensor sampling_with_replacement_cpu(
+  const Tensor& x,
+  const Tensor& weights,
+  int64_t k
+){
+  int n = x.size(0);
+  Tensor samples;
+
+  if (weights.numel() == 0){
+    samples = at::randint(0, n, {k}, x.options().dtype(at::kLong));
+  } else {
+
+    THGenerator* generator = get_generator(nullptr);
+    samples = at::empty({k}, x.options().dtype(at::kLong));
+    int64_t *samples_ptr = samples.data<int64_t>();
+
+    Tensor cdf = weights.cumsum(0);
+    cdf /= cdf[-1];
+
+    AT_DISPATCH_FLOATING_TYPES(weights.scalar_type(), "Sampling with replacement", [&] {
+      scalar_t *cdf_ptr = cdf.data<scalar_t>();
+      for(int i = 0; i < k; i++){
+        scalar_t u = THRandom_standard_uniform(generator);
+        auto ptr = std::lower_bound(cdf_ptr, cdf_ptr + n, u);
+        samples_ptr[i] = std::distance(cdf_ptr, ptr);
+      }
+    });
+
+  }
+
+  return x.index_select(0, samples);
+}
+
+Tensor choice_cpu(
+  const Tensor& input,
+  const Tensor& weights,
+  bool replace,
+  int64_t k
+){
+  if (replace){
+    return sampling_with_replacement_cpu(input, weights, k);
+  } else {
+    return reservoir_sampling_cpu(input, weights, k);
+  }
+}
+
+Tensor choice_cpu(
+  const Tensor& input,
+  bool replace,
+  int64_t k
+){
+  at::Tensor weights = at::empty({0}, input.options().dtype(at::kFloat));
+  if (replace){
```

Comment:
Here instead of duplicating the code you can just call through to `native::choice_cpu(input, weights, replace, k)`

Reply:
Indeed, I'll make this change.